### PR TITLE
fix(redblack): fix red/black deploy

### PR DESCRIPTION
**Pull request by marchello2000**
The main issue being resolved here is how we determine if a source ASG is present during R/B deploy.
In some cases, the `.asgName` property is used while others `.serverGroupName` is used. If we don't check both, some UI (tasks) operations don't work correctly,
e.g. clone with red/black and leave max 2 remaining ASGs will fail to disable the old ASG as well skip the shrink step.

While here, I am also making similar change to RRB and MD strategies.

Additionally, cleaning up the `composeFlow` -> `composeBefore/AfterFlow` for strategies as a follow up from PR #3133

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RedBlackStrategy.groovy [hunk 1]
```diff
@@ -53,38 +53,25 @@ class RedBlackStrategy implements Strategy, ApplicationContextAware {
...
-      // deprecated aroundStages
-      return []
-    }
 
     StageData parentStageData = parentCreateServerGroupStage.mapTo(StageData)
     StageData.Source sourceServerGroup = parentStageData.source
 
     // Short-circuit if there is no source server group
-    if (sourceServerGroup == null || sourceServerGroup.serverGroupName == null) {
+    if (sourceServerGroup == null || (sourceServerGroup.serverGroupName == null && sourceServerGroup.asgName == null)) {
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RedBlackStrategy.groovy]: this is the actual fix for the issue of not being able to perform a correct red/black clone on aws from the UI

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy [hunk 2]
```diff
@@ -74,18 +75,28 @@ class MonitoredDeployStrategy implements Strategy {
   DeploymentMonitorServiceProvider deploymentMonitorServiceProvider
 
   @Override
-  List<Stage> composeBeforeStages(Stage parent) {
-    return composeFlow(parent)
-      .findAll({ it.syntheticStageOwner == SyntheticStageOwner.STAGE_BEFORE })
-  }
+  List<Stage> composeBeforeStages(Stage stage) {
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy]: this is annoying, but some strategies (like RRB) took advantage of being called during `plan` and mutating the current stage. Hence this code lives in `composeBeforeStages` which is also called during stage `plan`ning

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy [hunk 3]
```diff
@@ -132,23 +131,16 @@ class MonitoredDeployStrategy implements Strategy {
...
-
-      if (parentCreateServerGroupStage.status == ExecutionStatus.NOT_STARTED) {
-        // No point in composing the flow if we are called to plan "beforeStages" since we don't have any STAGE_BEFOREs.
-        // Also, we rely on the the source server group task to have run already.
-        // In the near future we will move composeFlow into beforeStages and afterStages instead of the
-        // deprecated aroundStages
-        return []
-      }
+      Stage parentCreateServerGroupStage = stage.directAncestors()
+        .find() { it.type == CreateServerGroupStage.PIPELINE_CONFIG_TYPE || it.type == CloneServerGroupStage.PIPELINE_CONFIG_TYPE }
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/MonitoredDeployStrategy.groovy]: this allows the strategy to be used for `clone` as well as `create` servergroup

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy [hunk 4]
```diff
@@ -124,23 +123,16 @@ class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
...
-
-      if (parentCreateServerGroupStage.status == ExecutionStatus.NOT_STARTED) {
-        // No point in composing the flow if we are called to plan "beforeStages" since we don't have any STAGE_BEFOREs.
-        // Also, we rely on the the source server group task to have run already.
-        // In the near future we will move composeFlow into beforeStages and afterStages instead of the
-        // deprecated aroundStages
-        return []
-      }
+      Stage parentCreateServerGroupStage = stage.directAncestors()
+        .find() { it.type == CreateServerGroupStage.PIPELINE_CONFIG_TYPE || it.type == CloneServerGroupStage.PIPELINE_CONFIG_TYPE }
```
Comment (marchello2000) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy]: this allows the strategy to be used for `clone` as well as `create` servergroup

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RedBlackStrategy.groovy (hunk 1 above)
Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RedBlackStrategy.groovy]: I am flabbergasted, nice find 🥇

**Pull request merged**